### PR TITLE
netplan status --diff fixes and improvements

### PR DESCRIPTION
**Pull request by daniloegea**
This PR contains a bunch of small improvements to `netplan status --diff` and some additions to the libnetplan API:

- Add getters for the `gateway4` and `gateway6` properties. This is required during the analysis of routes.
- Routes from `gateway4` and `gateway6` are added to the list of routes being analyzed.
- Make use of the property `dynamic` from `addr_info` to help determining if the address was received dynamically.
- Make use of the `ra` flag from routes to help finding IPs that were received via RA.
- Add getter for the `accept-ra` property.
- Improve the analysis of RA and LL addresses.
- Fix a rendering issue in the `--diff` mode when `python3-rich` is not present.
- Use the ConfigSource property from networkd to reduce the guessing and be more precise when analyzing dynamic configuration.

I recommend testing it inside a LXD VM where IPV6 addresses are assigned dynamically. The currently netplan.io will show the IPv6 address and DNS as diffs. The heuristics added to the analysis help to reduce this noise.

## Checklist

- [ ] Runs `make check` successfully.
- [ ] Retains 100% code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

### CODE REVIEW

File: netplan_cli/cli/state_diff.py [hunk 1]
```diff
@@ -442,6 +486,62 @@ def _normalize_routes(self, routes: set) -> set:
...
+        interface = config.get('id')
+
+        ifaces = [iface for iface in system_interfaces.values()
+                  if iface.get('id') == interface]
+        routes = [route for iface in ifaces for route in iface.get('routes', [])]
+
+        if gateway4 := config.get('gateway4'):
+            default_routes = [route for route in routes
+                              if route.get('to') == 'default'
+                              and route.get('family') == 2
```
Comment (slyon) [inline netplan_cli/cli/state_diff.py:511]: nitpick: use `socket.AF_INET` named constant, also in the `route = NetplanRoute(...)` below.

And same for `socket.AF_INET6` in `gateway6` below.

File: netplan_cli/cli/state_diff.py [hunk 2]
```diff
@@ -442,6 +486,62 @@ def _normalize_routes(self, routes: set) -> set:
...
+        system_interfaces = self.system_state.get_data()
+        interface = config.get('id')
+
+        ifaces = [iface for iface in system_interfaces.values()
+                  if iface.get('id') == interface]
+        routes = [route for iface in ifaces for route in iface.get('routes', [])]
+
+        if gateway4 := config.get('gateway4'):
+            default_routes = [route for route in routes
+                              if route.get('to') == 'default'
```
Comment (slyon) [inline netplan_cli/cli/state_diff.py:510]: question (non-blocking): Are we sure this is already normalized to `default`, or could it also be something like `0.0.0.0/0`?

File: netplan_cli/cli/state_diff.py [hunk 3]
```diff
@@ -442,6 +486,62 @@ def _normalize_routes(self, routes: set) -> set:
 
         return new_routes_set
 
+    def _normalize_gateways(self, config: dict) -> AbstractSet[NetplanRoute]:
```
Comment (slyon) [inline netplan_cli/cli/state_diff.py:489]: thought: Wow, that's lots of code/logic for those deprecated `gateway4/6` fields.. But now that we have it we should keep it ;-)

File: netplan_cli/cli/state_diff.py [hunk 4]
```diff
@@ -486,7 +591,7 @@ def _filter_system_routes(self, system_routes: AbstractSet[NetplanRoute], system
                 continue
 
             # Filter IPv6 local routes
-            if route.family == 10 and (route.to in local_networks or route.to in addresses):
+            if route.family == 10 and route.protocol != 'ra' and (route.to in local_networks or route.to in addresses):
```
Comment (slyon) [inline netplan_cli/cli/state_diff.py:594]: Please use `socket.AF_INET6` instead.

File: tests/cli/test_state_diff.py [hunk 5]
```diff
@@ -628,6 +628,326 @@ def test_diff_dhcp_addresses_are_filtered_out(self):
...
+                        "from": "10.86.126.253",
+                        "metric": 100,
+                        "type": "unicast",
+                        "scope": "global",
+                        "protocol": "dhcp",
+                        "table": "main"
+                    },
+                    {
+                        "to": "10.86.126.0/24",
+                        "family": 2,
```
Comment (slyon) [inline tests/cli/test_state_diff.py:671]: nitpick: Consider using proper constants here and in the other tests below.

**Review by slyon — COMMENTED** [review]
Thank you very much for those enhancements and quality of life improvements!

Looks mostly good to me. Please consider using named constants for the ip-address family instead of plain numbers.

Also, we should avoid introducing new public API for deprecated symbols (gateay4/6), still handling them as you did via internal/private API is fine!

File: tests/cli/test_state_diff.py (hunk 5 above)
Comment (daniloegea) [reply tests/cli/test_state_diff.py:671]: hm I think I'll leave these as numbers. This is copied from state dumps and they will be represented as numbers. But I totally agree we should use the constants in non-test code.

**Comment by daniloegea** [comment]
Thank you, Lukas. I addressed your comments and tried to improve a little the analysis of dynamic configuration with networkd.

File: include/netdef.h [hunk 6]
```diff
@@ -217,3 +227,5 @@ netplan_netdef_get_link_local_ipv6(const NetplanNetDefinition* netdef);
  */
 NETPLAN_PUBLIC ssize_t
 netplan_netdef_get_macaddress(const NetplanNetDefinition* netdef, char* out_buffer, size_t out_buffer_size);
+
+
```
Comment (slyon) [inline include/netdef.h:231]: nitpick (non-blocking): Unnecessary newlines.

File: tests/cli/test_state_diff.py (hunk 5 above)
Comment (slyon) [reply tests/cli/test_state_diff.py:678]: fair enough!

**Review by slyon — APPROVED** [review]
Thanks a lot for addressing my remarks! I like the idea of using networkd's `ConfigSource`. Handling similar cases through NetworkManager's DBus API in the future is a good approach!

I pushed a tiny commit to clean up some unnecessary newlines/whitespace in `include/netdef.h`. Otherwise, LGTM. Let's get this merged!

File: include/netdef.h (hunk 6 above)
Comment (slyon) [reply include/netdef.h]: I fixed this in https://github.com/canonical/netplan/pull/466/commits/692f9023432ea35cc5e36485e7c5dd12ecd61b96

**Pull request merged**